search/includes/classes/class-versioning.php

```diff
@@ -745,4 +755,128 @@ public function action__ep_delete_indexable( $object_id, $indexable_slug ) {
 		// Clear the flag to return to normal
 		$this->is_doing_object_delete = false;
 	}
+
+	/**
+	 * Check if the versions are persisted correctly. If not recreate them.
+	 */
+	public function maybe_self_heal() {
+		if ( $this->is_versioning_valid() ) {
+			return;
+		}
+
```

Comment:
Needs memcache lock + handle requests while healing.

Any tips on how to do those things?

Reply:
For the memcache lock, the idea is to prevent multiple instances of this "repair" routine from running in parallel, which can quickly spiral out of control on high traffic sites.

The easiest way to do this is to just set a cache entry like `wp_cache_set( $key, 1, $group, $ttl )` when the routine starts, and check if it's already locked prior to starting, with `1 === wp_cache_get( $key )`. Oh and we should clean it up after the routine finishes, in case it gets corrupted again - `wp_cache_delete( $key, $group )`.

For handling requests during this...that's a tricky one. One one hand, we could check this key prior to every request and just tell it to fall back to the DB...but at the same time, adding the overhead of a memcache roundtrip to each request that hits ES isn't ideal, especially since this case will be super rare. The healing routine should also be pretty fast, so special handling for it might be an overoptimization that we can skip for now.

Reply:
Great thanks for the directions! I put it to dedicated methods to hopefully be more readable.